src/datachain/lib/dc.py

```diff
@@ -331,22 +346,80 @@ def from_storage(
             chain = DataChain.from_storage("s3://my-bucket/my-dir")
             ```
         """
-        func = get_file(type)
-        return (
-            cls(
-                path,
+        file_type = get_file_type(type)
+
+        if client_config is None:
+            client_config = {}
+
+        if anon:
+            client_config["anon"] = True
+
+        session = Session.get(
+            session,
+            catalog=kwargs.get("catalog"),
+            client_config=client_config,
+            in_memory=in_memory,
+        )
+
+        client, path = Client.parse_url(
+            uri, session.catalog.cache, **session.catalog.client_config
+        )
+
+        # clean path without globs
+        lst_path = (
```

Comment:
is it possible to extract all the logic to lib/listing.py? dc.py is becoming an unmanageble fat file

Reply:
I've tried but code looks worse than here. I've refactored it and reduced line numbers by ~half and there is not much logic left anyway except check if listing exist and if not call listing generator to create it, so I think it's good now.